### DISCUSSION TITLE
-x option with commands in a pipe

### DISCUSSION THREAD

**Discussion by ivan94fi** [sharkdp/fd#1133] [General]
Hi! Thanks for the awesome utility, it is really a game changer.

I was wondering if I could do something along the lines of
```shell
fd <options> -x 'command1 {} | command2'
```

Specifically, I would like to count the number of files in each directory in the current folder. I tried some alternatives:
```shell
fd -t d -d 1 -x 'ls {} | wc -l'  # not what I want: the entire string is treated as a command

fd -t d -d 1 -x ls {} \| wc -l  # not what I want: the escaped pipe, "wc" and "-l" are given in input to ls

fd -t d -d 1 -x ls {} | wc -l  # not what I want: the entire output of fd is piped in wc
```

This can be done with a for loop, but it would be good to avoid this:
```shell
for d in $(fd -t d -d 1); do ls $d | wc -l; done
```

**Comment by tavianator** [comment]
Something like `-x sh -c 'ls -1 "$1" | wc -l' sh {}` will work.

**Reply by ivan94fi** [reply]
Hi!

Thanks, this works as expected.

**Comment by ivan94fi** [comment]
**EDIT**: see comment below on differences between this version and the original version by @tavianator.

---

Hi, a small update on this. I figured a simpler way to execute commands with pipes ~~, avoiding the double `sh` invocation~~. I tested it on a directory with spaces in subdirectory names and it seems to work the same as @tavianator answer.

Here is the reviewed command:
```shell
fd -x sh -c 'ls "{}" | wc -l'
```

The `-1` argument to `ls` can be dropped as it is the default behaviour of `ls` when the output file is a pipe. ~~Also, the second shell invocation is removed.~~

~~I think this version is equivalent to the previous, but I may miss some edge cases.~~
This version definitely misses some edge cases.

P.S.
This also works with more complex commands: I personally used this to extract selected infos from videos, using a combination of mediainfo and jq:
```shell
fd -x sh -c "mediainfo --Output=JSON {} | jq '.media.track | .[] | select(.\"@type\" == \"Video\") | .Height'"
```

**Reply by tavianator** [reply]
My command didn't have a double `sh` invocation.  The second `sh` is just the zeroth argument to the shell command, it could be `foo` or something instead.

The reason to do `$1` instead of `{}` is in case you have a file named `"; sudo rm -rf /; : "`.  My version is safe, but yours will potentially execute commands from files with strange names.

**Reply by ivan94fi** [reply]
Perfect! I knew I was missing something! Thanks again
